CASSGO-82 CASSGO-83 fix tinyint decode bug and vector decode bug with SliceMap

joao-r-reis opened the pull request:
CASSGO-82

Tinyint unmarshal silently errors and unmarshals tinyint values as 0 in some situations.

This PR also removes other cases of silent errors in decode functions.


CASSGO-83

fix issue where reading vectors with SliceMap is not possible

jameshartig commented on marshal.go:
```diff
@@ -2352,7 +2370,7 @@ func (inetType) Type() Type {
 // Zero returns the zero value for the inet CQL type.
 func (inetType) Zero() interface{} {
-	return net.IP(nil)
+	return net.IP{}
```
What's this change for? net.IP is a slice and the zero value for a slice is nil.

joao-r-reis replied:
I went back and forth on this and I forgot to revert back to `nil` good catch

joao-r-reis replied:
should be fixed now

jameshartig commented on vector.go:
```diff
@@ -173,7 +183,7 @@ func (v VectorType) Unmarshal(data []byte, value interface{}) error {
 		return nil
 	}
-	return unmarshalErrorf("can not unmarshal %s into %T. Accepted types: slice, array.", v, value)
+	return unmarshalErrorf("can not unmarshal %s into %T. Accepted types: slice, array, *interface{}.", v, value)
```
this should just be `interface{}` or we should make the others be `*slice`, `*array`

joao-r-reis replied:
done

jameshartig commented on marshal.go:
```diff
@@ -636,26 +644,30 @@ func (intTypeInfo) Marshal(value interface{}) ([]byte, error) {
+func decInt(x []byte) (int32, error) {
 	if len(x) != 4 {
-		return 0
+		return 0, fmt.Errorf("expected 4 bytes decoding int but got %v", len(x))
```
I'm confused, isn't this going to make `NULL` now return an error instead of `0` like it did before?

jameshartig replied:
For MapScan/SliceMap they end up never leaving the Unmarshal method because we handle `*interface{}` and `nil` data directly in there. But I'm worried now if you Unmarshal with a `*int8` and a NULL it's going to error rather than set the value to `0`.

joao-r-reis replied:
> I'm confused, isn't this going to make NULL now return an error instead of 0 like it did before?

Yeah I started seeing a couple of test failures and I noticed that, was just working on it now. Should be fixed now.

jameshartig commented on vector.go:
```diff
@@ -143,6 +150,9 @@ func (v VectorType) Unmarshal(data []byte, value interface{}) error {
 			}
 		} else {
 			rv.Set(reflect.MakeSlice(t, v.Dimensions, v.Dimensions))
+			if rv.Kind() == reflect.Interface {
```
I don't think this can ever be true because you just set `rv` to a slice.

joao-r-reis replied:
It can be a variable of type `interface{}` that just got assigned a slice no?

jameshartig commented on vector.go:
```diff
@@ -124,6 +124,13 @@ func (v VectorType) Unmarshal(data []byte, value interface{}) error {
+		if t.NumMethod() != 0 {
+			return unmarshalErrorf("can not unmarshal into non-empty interface %T", value)
+		}
+		t = reflect.TypeOf(v.Zero())
```
This is assuming that the interface holds that type but it might not, I think you need to instead check if it's nil and then if its not nil, then do `rv = rv.Elem() t = rv.Type()`

joao-r-reis committed:
Fix tinyint silent unmarshal error and vector SliceMap bug

Tinyint unmarshal silently errors and unmarshals tinyint values as 0 in some situations.
This patch also removes other cases of silent errors in decode functions.

Fix issue where reading vectors with SliceMap is not possible.

Patch by João Reis; reviewed by James Hartig for CASSGO-82, CASSGO-83
